testcases/kernel/syscalls/madvise/madvise08.c

```diff
@@ -0,0 +1,297 @@
+/*
+ * Check that memory marked with MADV_DONTDUMP is not included in a core dump
+ * and check that the same memory then marked with MADV_DODUMP is included in
+ * a core dump.
+ *
+ * In order to reliably find the core dump this test temporarily changes the
+ * system wide core_pattern setting. Meaning all core dumps will be sent to the
+ * test's temporary dir untill the setting is restored during cleanup.
+ *
+ * Test flow: map memory,
+ *	      write generated character sequence to memory,
+ *	      start child process,
+ *	      mark memory with MADV_DONTDUMP in child,
+ *	      abort child,
+ *	      scan child's core dump for character sequence,
+ *	      if the sequence is not found it is a pass otherwise a fail,
+ *	      do the opposite for MADV_DODUMP,
+ */
+
+#include "tst_test.h"
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#if defined(MADV_DONTDUMP) && defined(MADV_DODUMP)
+#define CORE_PATTERN "/proc/sys/kernel/core_pattern"
+
+struct test_case {
+	int fd;
+	char *fname;
+	char *fcontent;
+	void *fmem;
+	size_t fmemsize;
+	int ycount;
+	char cpattern[1048];
+	ssize_t cpsize;
+	int cpfd;
+	struct stat stat;
+	int dfd;
+} tc;
+
+/*
+ *TODO: Check/set /proc/PID/coredump_filter,
+ *		RLIMIT_CORE, RLIMIT_FSIZE,
+ *		PR_SET_DUMPABLE,
+ *		CONFIG_COREDUMP (kernel)
+ * Currently it is assumed that if the core dump can not be found it is because
+ * they are disabled rather than a bug. Also tests may pass because of the core
+ * and file size limits.
+ */
+
+static void setup(void)
+{
+	const char *dumpf = "dump-%p";
+	char cwd[1024];
+	char tmpcpattern[1048];
+
+	tc.fd = 0;
+	tc.fname = "mmf0";
+	tc.fmem = NULL;
+	tc.cpsize = -1;
+	tc.cpfd = 0;
+	tc.dfd = 0;
+	tc.ycount = 1280;
+	tc.fmemsize = tc.ycount + 2;
+
+	SAFE_GETCWD(cwd, 1024);
+	snprintf(tmpcpattern, 1048, "%s/%s", cwd, dumpf);
+	tst_res(TINFO, "Temporary core pattern is '%s'", tmpcpattern);
+
+	tc.cpfd = SAFE_OPEN(CORE_PATTERN, O_RDONLY);
+	tc.cpsize = SAFE_READ(0, tc.cpfd, &tc.cpattern, 1048);
+	SAFE_CLOSE(tc.cpfd);
+	tc.cpfd = 0;
+
+	if (tc.cpsize == 1048)
+		tst_brk(TCONF, "Current core pattern is too large to save");
+
+	tc.cpattern[tc.cpsize + 1] = '\0';
+	tst_res(TINFO, "System core pattern is '%s'", tc.cpattern);
+
+	tc.cpfd = SAFE_OPEN(CORE_PATTERN, O_TRUNC | O_WRONLY);
+	SAFE_WRITE(1, tc.cpfd, tmpcpattern, strlen(tmpcpattern));
+	SAFE_CLOSE(tc.cpfd);
+	tc.cpfd = 0;
+}
+
+static void restore_cpattern(void)
+{
+	int fd, rval;
+	size_t written;
+
+	if (tc.cpfd != 0 && close(tc.cpfd))
+		tst_res(TINFO | TERRNO, "close(tc.cpdf)");
+
+	if (tc.cpsize < 1)
+		return;
+
+	fd = open(CORE_PATTERN, O_TRUNC | O_WRONLY);
+	if (fd == -1) {
+		tst_res(TINFO | TERRNO, "open(CORE_PATTERN, O_TRUNC)");
+		goto error;
+	}
+
+	written = write(fd, tc.cpattern, tc.cpsize);
+	if (written < (size_t)tc.cpsize) {
+		tst_brk(TINFO | TERRNO, "write(fd, tc.pattern, ...)");
+		goto error;
+	}
```

Comment:
Are you aware that tst_brk() will exit the test execution immediately?

All you need to do to restore the pattern is to call ```FILE_PRINTF()``` from the ```cleanup()``` since the ```SAFE_``` version will abort the ```cleanup()``` on a failure. And it looks like it's not exported to the new library yet. I will fix that.